src/main/java/foundation/privacybydesign/sms/TokenManager.java

```diff
@@ -96,28 +109,161 @@ public static boolean isEqualsConstantTime(char[] a, char[] b) {
-            }
-        }
+        tokenRepo.removeExpired();
     }
 }
 
 class TokenRequest {
     String token;
     int tries;
-    private long created;
+    long created;
+
+    TokenRequest(String token, int tries, long created) {
+        this.token = token;
+        this.tries = tries;
+        this.created = created;
+    }
 
     TokenRequest(String token) {
         this.token = token;
         created = System.currentTimeMillis();
         tries = 0;
     }
 
+    static boolean isExpiredForCreationDate(long creationDate) {
+        return System.currentTimeMillis() - creationDate > SMSConfiguration.getInstance().getSMSTokenValidity() * 1000;
+    }
+
     boolean isExpired() {
-        return System.currentTimeMillis() - this.created >
-                SMSConfiguration.getInstance().getSMSTokenValidity()*1000;
+        return isExpiredForCreationDate(created);
+    }
+}
+
+/**
+ * An interface for a repository to store tokens in.
+ */
+interface TokenRequestRepository {
+    void store(String phone, TokenRequest request);
+
+    TokenRequest retrieve(String phone);
+
+    void remove(String phone);
+
+    void removeExpired();
+}
+
+/**
+ * A token repository that stores and retrieves tokens from a redis store.
+ * Useful for when the sms issuer needs to be stateless.
+ */
+class RedisTokenRequestRepository implements TokenRequestRepository {
+    private static Logger logger = LoggerFactory.getLogger(RedisTokenRequestRepository.class);
+    private static String namespace = "request";
+    JedisSentinelPool pool;
+
+    RedisTokenRequestRepository() {
+        pool = Redis.createSentinelPoolFromEnv();
+    }
+
+    @Override
+    public void store(String phone, TokenRequest request) {
+        final String key = Redis.createKey(namespace, phone);
+        try (var jedis = pool.getResource()) {
+            jedis.hset(key, "token", request.token);
+            jedis.hset(key, "tries", Integer.toString(request.tries));
+            jedis.hset(key, "created", String.valueOf(request.created));
+        }
+    }
+
+    @Override
+    public void remove(String phone) {
+        final String key = Redis.createKey(namespace, phone);
+
+        try (var jedis = pool.getResource()) {
+            jedis.del(key);
+        }
+    }
+
+    @Override
+    public void removeExpired() {
+        final String pattern = Redis.createNamespace(namespace) + "*";
+        ScanParams scanParams = new ScanParams().match(pattern);
+        String cursor = "0";
+
+        try (var jedis = pool.getResource()) {
+            do {
+                ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
+                List<String> keys = scanResult.getResult();
+                cursor = scanResult.getCursor();
+
+                for (String key : keys) {
+                    removeIfExpired(jedis, key);
+                }
+            } while (!cursor.equals("0")); // continue until the cursor wraps around
+        }
+    }
```

Comment:
This would be better, but for now we'll just stick with the code that was already there as much as possible in order not to break anything. I've added todo comments, so this is noted in the code.